llvm/test/Transforms/InstCombine/fsqrtdiv-transform.ll

```diff
@@ -0,0 +1,459 @@
+  ret void
+}
+
+; missing flags for optimization.
+define void @missing_flags_on_mul(double %a) {
+; CHECK-LABEL: define void @missing_flags_on_mul(
+; CHECK-SAME: double [[A:%.*]]) {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[SQRT:%.*]] = tail call double @llvm.sqrt.f64(double [[A]])
+; CHECK-NEXT:    [[DIV:%.*]] = fdiv reassoc arcp double 1.000000e+00, [[SQRT]]
+; CHECK-NEXT:    store double [[DIV]], ptr @x, align 8
+; CHECK-NEXT:    [[MUL:%.*]] = fmul double [[DIV]], [[DIV]]
+; CHECK-NEXT:    store double [[MUL]], ptr @r1, align 8
+; CHECK-NEXT:    [[DIV1:%.*]] = fdiv reassoc nsz double [[A]], [[SQRT]]
+; CHECK-NEXT:    store double [[DIV1]], ptr @r2, align 8
+; CHECK-NEXT:    ret void
+entry:
+  %sqrt = tail call double @llvm.sqrt.f64(double %a)
+  %div = fdiv reassoc arcp double 1.000000e+00, %sqrt
+  store double %div, ptr @x
+  %mul = fmul double %div, %div
+  store double %mul, ptr @r1
+  %div1 = fdiv reassoc nsz double %a, %sqrt
+  store double %div1, ptr @r2
+  ret void
+}
+
+; missing flags for optimization.
+define void @missing_flags_on_div1(double %a) {
+; CHECK-LABEL: define void @missing_flags_on_div1(
+; CHECK-SAME: double [[A:%.*]]) {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[SQRT:%.*]] = tail call double @llvm.sqrt.f64(double [[A]])
+; CHECK-NEXT:    [[DIV:%.*]] = fdiv reassoc arcp double 1.000000e+00, [[SQRT]]
+; CHECK-NEXT:    store double [[DIV]], ptr @x, align 8
+; CHECK-NEXT:    [[MUL:%.*]] = fmul reassoc nnan nsz double [[DIV]], [[DIV]]
+; CHECK-NEXT:    store double [[MUL]], ptr @r1, align 8
+; CHECK-NEXT:    [[DIV1:%.*]] = fdiv double [[A]], [[SQRT]]
+; CHECK-NEXT:    store double [[DIV1]], ptr @r2, align 8
+; CHECK-NEXT:    ret void
+entry:
+  %sqrt = tail call double @llvm.sqrt.f64(double %a)
+  %div = fdiv reassoc arcp double 1.000000e+00, %sqrt
+  store double %div, ptr @x
+  %mul = fmul reassoc nnan nsz double %div, %div
+  store double %mul, ptr @r1
+  %div1 = fdiv double %a, %sqrt
+  store double %div1, ptr @r2
+  ret void
+}
+
+; div = -1/sqrt(a)
+define void @negative_fdiv_val(double %a) {
+; CHECK-LABEL: define void @negative_fdiv_val(
+; CHECK-SAME: double [[A:%.*]]) {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[SQRT:%.*]] = tail call double @llvm.sqrt.f64(double [[A]])
+; CHECK-NEXT:    [[TMP0:%.*]] = fdiv reassoc nnan nsz double 1.000000e+00, [[A]]
+; CHECK-NEXT:    [[TMP1:%.*]] = fneg reassoc arcp double [[TMP0]]
+; CHECK-NEXT:    [[DIV:%.*]] = fmul reassoc arcp double [[SQRT]], [[TMP1]]
+; CHECK-NEXT:    store double [[DIV]], ptr @x, align 8
+; CHECK-NEXT:    store double [[TMP0]], ptr @r1, align 8
+; CHECK-NEXT:    store double [[SQRT]], ptr @r2, align 8
+; CHECK-NEXT:    ret void
+entry:
+  %sqrt = tail call double @llvm.sqrt.f64(double %a)
+  %div = fdiv reassoc arcp double -1.000000e+00, %sqrt
+  store double %div, ptr @x
+  %mul = fmul reassoc nnan nsz double %div, %div
+  store double %mul, ptr @r1
+  %div1 = fdiv reassoc nsz double %a, %sqrt
+  store double %div1, ptr @r2
+  ret void
```

Comment:
Staring hard in this case, the transformation does not seem correct to me, though I don't have precise rationale for all the necessary preconditions without more work. If you called this expression with `%a = - 0.0`, then the results that we have are

```
  %sqrt = -0.0
  *@x = %div = -1.0 / -0.0 = +infinity
  *@r1 = %mul =  +inf * +inf = +inf
  *@r2 = %div1 = -0.0 / -0.0 [nsz] = nan
```

The transformed code is:
```
  %sqrt = -0.0
  *@r1 = 1.0 / -0.0 [nsz] = +/- infinity
  *@r2 = %sqrt = -0.0
  *@x = -0.0 * %tmp1 = -0.0 * -( +/- infinity) = nan
```

Reply:
@jcranmer-intel The simplest solution to me is having `nsz nnan ninf` on the sqrt call itself and remove the flags, except reassoc, on all the other instructions so that all cases are handled.